mesonbuild/build.py

```diff
@@ -967,16 +967,32 @@ def process_kwargs(self, kwargs, environment):
                 raise InvalidArguments('GNU symbol visibility arg %s not one of: %s',
                                        self.symbol_visibility, ', '.join(permitted))
 
-    def _extract_pic_pie(self, kwargs, arg):
+    def _extract_pic_pie(self, kwargs, arg: str) -> T.Optional[bool]:
         # Check if we have -fPIC, -fpic, -fPIE, or -fpie in cflags
         all_flags = self.extra_args['c'] + self.extra_args['cpp']
         if '-f' + arg.lower() in all_flags or '-f' + arg.upper() in all_flags:
             mlog.warning("Use the '{}' kwarg instead of passing '{}' manually to {!r}".format(arg, '-f' + arg, self.name))
             return True
+        if '-fno-' + arg.lower() in all_flags or '-fno-' + arg.upper() in all_flags:
+            mlog.warning("Use the '{}' kwarg instead of passing '{}' manually to {!r}".format(arg, '-fno-' + arg, self.name))
+            return False
 
-        val = kwargs.get(arg, False)
-        if not isinstance(val, bool):
-            raise InvalidArguments('Argument {} to {!r} must be boolean'.format(arg, self.name))
+        if arg == 'pic':
+            val = kwargs.get(arg, False)
+            if not isinstance(val, bool):
+                raise InvalidArguments(
+                    'Argument {} to {!r} must be a boolean'.format(arg, self.name))
+        else:
+            val = kwargs.get(arg, None)
+            if val == "default":
+                val = None
+            elif val == 'true':
+                val = True
+            elif val == 'false':
```

Comment:
still can't be a string `'true'` or `'false'` here. I think what you mean is:
```
val = kwargs.get(arg, None)
if val == 'default':
  val = None
elif not isinstance(val, bool):
  raise ...

```

Reply:
If the option comes from the `-Db_pie=` command line option it will be a string, because that's a combo.

Reply:
I don't think it can be string, otherwise current code wouldn't work. In Executable class it does: `kwargs['pie'] = environment.coredata.base_options['b_pie'].value`, and UserBooleanOption.value is a bool not string.

Reply:
It's not a UserBooleanOption in this series, it's a UserComboOption

Reply:
oohh, indeed, sorry completely missed the point here :D But that means that I would write `executable(..., pie : 'true')`, it's nitpicking but would be nice to forbid that syntax, no?

Reply:
Yeah, That probably should be done in the interpreter. I'll make some changes

Reply:
Okay, I've added an interpreter level check for strings other that `'default'` being passed to the `pie :` keyword argument. I've also added a FeatureNew for the default.

Reply:
Ok, I think that close this discussion here.